oishiline.3
.TH OISHILINE 3
.SH NAME
oishiline \- statusline and tabline for
.IR nvim (1).
.SH SYNOPSIS
.EX
.B require(\(dqoishiline\(dq).setup()
.EE
.SH DESCRIPTION
.B oishiline
is a statusline and tabline for
.IR nvim (1).
.SH DEFAULT SETUP CONFIGURATION
.EX
{
  tabline = { enable = true },

  globalArgs = {
    default = {
      fg = "#e5e9f0",
      bg = "#2e3440",
      ctermfg = 7,
      ctermbg = 0,
    },

    colors = {
      black = "#2e3440",
      darkred = "#bf616a",
      darkgreen = "#a3be8c",
      darkyellow = "#ebcb8b",
      darkblue = "#81a1c1",
      darkmagenta = "#b48ead",
      darkcyan = "#88c0d0",
      lightgray = "#e5e9f0",
      darkgray = "#4c566a",
      red = "#bf616a",
      green = "#a3be8c",
      yellow = "#ebcb8b",
      blue = "#81a1c1",
      magenta = "#b48ead",
      cyan = "#88c0d0",
      white = "#8fbcbb",
    },
  },

  statusline = {
    enable = true,

    leftModules = {
      { name = "mode" },
      { name = "separator" },
      { name = "branch" },
      { name = "filename" },
      { name = "diagnostics" },
    },

    rightModules = {
      {
        name = "separator",

        args = {
          suffix = "Right",

          active = {
            gui = "",
            tty = "",
          },

          inactive = {
            gui = "",
            tty = "",
          },

          dataHl = {
            fg = "#4c566a",
            bg = "#2e3440",
          },

          dataHlAlt = {
            fg = "#e5e9f0",
            bg = "#2e3440",
          },
        },
      },

      { name = "encoding" },
      { name = "format" },
      { name = "type" },

      {
        name = "separator",

        args = {
          suffix = "Right2",

          active = {
            gui = "",
            tty = "",
          },

          inactive = {
            gui = "",
            tty = "|",
          },

          dataHl = {
            fg = "#81a1c1",
            bg = "#4c566a",
            ctermfg = 4,
            ctermbg = 8,
          },

          dataHlAlt = {
            fg = "#e5e9f0",
            ctermfg = 7,
            ctermbg = 8,
          },
        },
      },

      { name = "progress" },
      { name = "location" },
    },
  },
}
.EE
.SH DEFAULT MODULE CONFIGURATION
.SS tabline
.EX
{
  suffix = "",

  dataHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  dataHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  iconHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  iconHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  leftPad = {
    gui = " ",
    tty = " ",
  },

  leftSep = {
    gui = "",
    tty = "",
  },

  leftSepHl = {
    bg = "#81a1c1",
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  leftSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 0,
    fg = "#2e3440",
  },

  rightPad = {
    gui = " ",
    tty = " ",
  },

  rightSep = {
    gui = "",
    tty = " ",
  },

  rightSepHl = {
    bg = "#2e3440",
    ctermbg = 0,
    ctermfg = 4,
    fg = "#81a1c1",
  },

  rightSepHlAlt = {
    bg = "#2e3440",
    ctermbg = 0,
    ctermfg = 8,
    fg = "#4c566a",
  },

  start = {
    gui = "",
    tty = "",
  },
}
.EE
.SS branch
.EX
{
  suffix = "",

  dataHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  dataHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  icon = {
    gui = " ",
    tty = "",
  },

  iconHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  iconHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  leftPad = {
    gui = " ",
    tty = " ",
  },

  leftSep = {
    gui = "",
    tty = "",
  },

  leftSepHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  leftSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 0,
    fg = "#2e3440",
  },

  rightPad = {
    gui = " ",
    tty = " ",
  },

  rightSep = {
    gui = "",
    tty = "|",
  },

  rightSepHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  rightSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },
}
.EE
.SS diagnostics
.EX
{
  suffix = "",

  sep = {
    gui = "",
    tty = "",
  },

  Error = {
    bg = "#bf616a",
    ctermbg = 1,
  },

  ErrorBg = {
    bg = "#bf616a",
    ctermbg = 1,
  },

  Warn = {
    bg = "#ebcb8b",
    ctermbg = 3,
  },

  WarnBg = {
    bg = "#ebcb8b",
    ctermbg = 3,
  },

  Info = {
    bg = "#81a1c1",
    ctermbg = 4,
  },

  InfoBg = {
    bg = "#81a1c1",
    ctermbg = 4,
  },

  Hint = {
    bg = "#88c0d0",
    ctermbg = 6,
  },

  HintBg = {
    bg = "#88c0d0",
    ctermbg = 6,
  },
}
.EE
.SS encoding
.EX
{
  suffix = "",

  dataHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  dataHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  icon = {
    gui = "",
    tty = "",
  },

  iconHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  iconHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  leftPad = {
    gui = " ",
    tty = " ",
  },

  leftSep = {
    gui = "",
    tty = "",
  },

  leftSepHl = {
    bg = "#2e3440",
    ctermbg = 0,
    ctermfg = 8,
    fg = "#4c566a",
  },

  leftSepHlAlt = {
    bg = "#2e3440",
    ctermbg = 0,
    ctermfg = 8,
    fg = "#4c566a",
  },

  rightPad = {
    gui = " ",
    tty = " ",
  },

  rightSep = {
    gui = "",
    tty = "|",
  },

  rightSepHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  rightSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },
}
.EE
.SS filename
.EX
{
  suffix = "",

  dataHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  dataHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  icon = {
    gui = "",
    tty = "",
  },

  iconHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  iconHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  leftPad = {
    gui = " ",
    tty = " ",
  },

  leftSep = {
    gui = "",
    tty = "",
  },

  leftSepHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  leftSepHlAlt = {
    bg = "#e5e9f0",
    ctermbg = 7,
    ctermfg = 0,
    fg = "#2e3440",
  },

  rightPad = {
    gui = " ",
    tty = " ",
  },

  rightSep = {
    gui = "",
    tty = " ",
  },

  rightSepHl = {
    bg = "#2e3440",
    ctermbg = 0,
    ctermfg = 1,
    fg = "#81a1c1",
  },

  rightSepHlAlt = {
    bg = "#2e3440",
    ctermbg = 0,
    ctermfg = 8,
    fg = "#4c566a",
  },
}
.EE
.SS format
.EX
{
  suffix = "",

  dataHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  dataHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  icon = {
    gui = "",
    tty = "",
  },

  iconHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  iconHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  leftPad = {
    gui = " ",
    tty = " ",
  },

  leftSep = {
    gui = "",
    tty = "",
  },

  leftSepHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  leftSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  rightPad = {
    gui = " ",
    tty = " ",
  },

  rightSep = {
    gui = "",
    tty = "",
  },

  rightSepHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  rightSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },
}
.EE
.SS location
.EX
{
  suffix = "",

  dataHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  dataHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  icon = {
    gui = "",
    tty = "",
  },

  iconHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  iconHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  leftPad = {
    gui = " ",
    tty = " ",
  },

  leftSep = {
    gui = "",
    tty = "",
  },

  leftSepHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  leftSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  rightPad = {
    gui = " ",
    tty = " ",
  },

  rightSep = {
    gui = "",
    tty = "",
  },

  rightSepHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  rightSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },
}
.EE
.SS mode
.EX
{
  suffix = "",

  command = {
    bg = "#81a1c1",
    ctermbg = 4,
  },

  insert = {
    bg = "#a3be8c",
    ctermbg = 2,
  },

  leftPad = {
    gui = " ",
    tty = " ",
  },

  leftSep = {
    gui = "",
    tty = "",
  },

  normal = {
    bg = "#81a1c1",
    ctermbg = 4,
  },

  replace = {
    bg = "#bf616a",
    ctermbg = 1,
  },

  rightPad = {
    gui = " ",
    tty = " ",
  },

  rightSep = {
    gui = "",
    tty = "",
  },

  terminal = {
    bg = "#a3be8c",
    ctermbg = 2,
  },

  visual = {
    bg = "#b48ead",
    ctermbg = 5,
  },
}
.EE
.SS progress
.EX
{
  suffix = "",

  dataHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  dataHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  icon = {
    gui = "",
    tty = "",
  },

  iconHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  iconHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  leftPad = {
    gui = " ",
    tty = " ",
  },

  leftSep = {
    gui = "",
    tty = "",
  },

  leftSepHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 4,
    fg = "#81a1c1",
  },

  leftSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 8,
    fg = "#4c566a",
  },

  rightPad = {
    gui = " ",
    tty = " ",
  },

  rightSep = {
    gui = "",
    tty = "|",
  },

  rightSepHl = {
    bg = "#81a1c1",
    bold = true,
    ctermbg = 4,
    ctermfg = 0,
    fg = "#2e3440",
  },

  rightSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },
}
.EE
.SS separator
.EX
{
  suffix = "",

  active = {
    gui = "",
    tty = " ",
  },

  dataHl = {
    bg = "#81a1c1",
    ctermbg = 0,
    ctermfg = 7,
    fg = "#2e3440",
  },

  dataHlAlt = {
    bg = "#4c566a",
    ctermbg = 0,
    ctermfg = 7,
    fg = "#2e3440",
  },

  inactive = {
    gui = "",
    tty = " ",
  },
}
.EE
.SS type
.EX
{
  suffix = "",

  dataHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  dataHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  icon = {
    gui = "",
    tty = "",
  },

  iconHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  iconHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  leftPad = {
    gui = " ",
    tty = " ",
  },

  leftSep = {
    gui = "",
    tty = "|",
  },

  leftSepHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  leftSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  rightPad = {
    gui = " ",
    tty = " ",
  },

  rightSep = {
    gui = "",
    tty = "",
  },

  rightSepHl = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },

  rightSepHlAlt = {
    bg = "#4c566a",
    ctermbg = 8,
    ctermfg = 7,
    fg = "#e5e9f0",
  },
}
.EE
.SH SEE ALSO
.IR nvim (1)
.SH AUTHOR
Kris Andrie Ortega (andrieee44@gmail.com)


lua/oishiline.lua
local lib = require("oishiline.modules.lib")
local M = {}

local function initArgs(userArgs)
	local colors = {
		black = "#2e3440",
		darkred = "#bf616a",
		darkgreen = "#a3be8c",
		darkyellow = "#ebcb8b",
		darkblue = "#81a1c1",
		darkmagenta = "#b48ead",
		darkcyan = "#88c0d0",
		lightgray = "#e5e9f0",
		darkgray = "#4c566a",
		red = "#bf616a",
		green = "#a3be8c",
		yellow = "#ebcb8b",
		blue = "#81a1c1",
		magenta = "#b48ead",
		cyan = "#88c0d0",
		white = "#8fbcbb",
	}

	local default = {
		fg = colors.lightgray,
		bg = colors.black,
		ctermfg = 7,
		ctermbg = 0,
	}

	local args = vim.tbl_deep_extend("keep", userArgs or {}, {
		tabline = { enable = true },

		globalArgs = {
			default = default,
			colors = colors,
		},

		statusline = {
			enable = true,

			leftModules = {
				{ name = "mode" },
				{ name = "separator" },
				{ name = "branch" },
				{ name = "filename" },
				{ name = "diagnostics" },
			},

			rightModules = {
				{
					name = "separator",

					args = {
						suffix = "Right",

						active = {
							gui = "",
							tty = "",
						},

						inactive = {
							gui = "",
							tty = "",
						},

						dataHl = {
							fg = colors.darkgray,
							bg = default.bg,
						},

						dataHlAlt = {
							fg = colors.darkgray,
							bg = default.bg,
						},
					},
				},

				{ name = "encoding" },
				{ name = "format" },
				{ name = "type" },

				{
					name = "separator",

					args = {
						suffix = "Right2",

						active = {
							gui = "",
							tty = "",
						},

						inactive = {
							gui = "",
							tty = "|",
						},

						dataHl = {
							fg = colors.darkblue,
							bg = colors.darkgray,
							ctermfg = 4,
							ctermbg = 8,
						},

						dataHlAlt = {
							fg = colors.lightgray,
							ctermfg = 7,
							ctermbg = 8,
						},
					},
				},

				{ name = "progress" },
				{ name = "location" },
			},
		},
	})

	vim.api.nvim_set_hl(0, "OishilineDefault", args.globalArgs.default)

	return args
end

local function initStatusline(args)
	local oishiline = vim.g.oishiline

	oishiline.statusline = {
		leftModules = {},
		rightModules = {},
	}

	for i, v in ipairs(args.statusline.leftModules) do
		oishiline.statusline.leftModules[i] = require(string.format("oishiline.modules.%s", v.name))()
		oishiline.statusline.leftModules[i].init(args.globalArgs, v.args or {})
	end

	for i, v in ipairs(args.statusline.rightModules) do
		oishiline.statusline.rightModules[i] = require(string.format("oishiline.modules.%s", v.name))()
		oishiline.statusline.rightModules[i].init(args.globalArgs, v.args or {})
	end

	vim.g.oishiline = oishiline
	vim.opt_global.statusline = "%{%v:lua.require('oishiline').statusline()%}"
end

local function initTabline(args)
	local oishiline = vim.g.oishiline
	oishiline.tabline = require("oishiline.tabline")
	oishiline.tabline.init(args.globalArgs, args.tabline)
	vim.g.oishiline = oishiline
	vim.opt_global.tabline = "%!v:lua.require('oishiline').tabline()"
end

function M.statusline()
	local results = {}
	local i = 1

	i = lib.insert(results, i, "%#OishilineDefault#")

	for _, v in ipairs(vim.g.oishiline.statusline.leftModules) do
		i = lib.insert(results, i, v.run())
	end

	i = lib.insert(results, i, "%=%(")

	for _, v in ipairs(vim.g.oishiline.statusline.rightModules) do
		i = lib.insert(results, i, v.run())
	end

	i = lib.insert(results, i, "%)")

	return table.concat(results)
end

function M.tabline()
	return vim.g.oishiline.tabline.run()
end

function M.setup(userArgs)
	local args = initArgs(userArgs)
	vim.g.oishiline = { statusline = {} }

	if args.statusline.enable then
		initStatusline(args)
	end

	if args.tabline.enable then
		initTabline(args)
	end
end

return M
